### PR TITLE
impl(GCS+gRPC): automatically refresh gRPC channels

### PR DESCRIPTION
**Pull request by coryan**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9905)

### CODE REVIEW

File: google/cloud/internal/async_connection_ready.h [hunk 1]
```diff
@@ -50,14 +51,30 @@ class AsyncConnectionReadyFuture
 
   // gRPC uses an anonymous type for the gRPC channel state enum :shrug:.
   using ChannelStateType = decltype(GRPC_CHANNEL_READY);
```
Comment (devbww) [inline google/cloud/internal/async_connection_ready.h:53]: Now unused (at least here)?

File: google/cloud/storage/internal/storage_round_robin.cc [hunk 2]
```diff
@@ -262,6 +281,32 @@ std::shared_ptr<StorageStub> StorageRoundRobin::Child() {
   return child;
 }
 
+void StorageRoundRobin::Refresh(
+    std::size_t index,
+    std::weak_ptr<google::cloud::internal::CompletionQueueImpl> wcq) {
+  auto cq = wcq.lock();
+  if (!cq) return;
+  auto deadline = std::chrono::system_clock::now() + kRefreshPeriod;
+  // An invalid index, stop the loop.  There is no need to check lock the mutex,
```
Comment (devbww) [inline google/cloud/storage/internal/storage_round_robin.cc:290]: s/check lock/lock/?

**Review by devbww — APPROVED** [review]

File: google/cloud/internal/async_connection_ready.h (hunk 1 above)
Comment (coryan) [reply google/cloud/internal/async_connection_ready.h]: Removed.

File: google/cloud/storage/internal/storage_round_robin.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/internal/storage_round_robin.cc]: Fixed.

File: google/cloud/internal/async_connection_ready.cc [hunk 3]
```diff
@@ -27,7 +27,8 @@ AsyncConnectionReadyFuture::AsyncConnectionReadyFuture(
     : cq_(std::move(cq)), channel_(std::move(channel)), deadline_(deadline) {}
 
 future<Status> AsyncConnectionReadyFuture::Start() {
-  RunIteration(channel_->GetState(true));
+  channel_->GetState(true);
```
Comment (dbolduc) [inline google/cloud/internal/async_connection_ready.cc:30]: nit: I think this call is superfluous

Comment (devbww) [reply google/cloud/internal/async_connection_ready.cc:30]: I looked at that too, but it causes the channel to try to make a connection if it is currently IDLE.

Comment (dbolduc) [reply google/cloud/internal/async_connection_ready.cc:30]: won't it get called here anyways? https://github.com/googleapis/google-cloud-cpp/blob/55f718928d4ff14515dde192763715c89a7746bf/google/cloud/internal/async_connection_ready.cc#L74

File: google/cloud/storage/internal/storage_round_robin.cc [hunk 4]
```diff
@@ -262,6 +281,32 @@ std::shared_ptr<StorageStub> StorageRoundRobin::Child() {
...
 }
 
+void StorageRoundRobin::Refresh(
+    std::size_t index,
+    std::weak_ptr<google::cloud::internal::CompletionQueueImpl> wcq) {
+  auto cq = wcq.lock();
+  if (!cq) return;
+  auto deadline = std::chrono::system_clock::now() + kRefreshPeriod;
+  // An invalid index, stop the loop.  There is no need to lock the mutex, as
+  // the channels do not change after the class is initialized.
```
Comment (devbww) [inline google/cloud/storage/internal/storage_round_robin.cc:291]: Q.  How is an invalid index injected in the first place?  If `channels_` doesn't change after initialization, couldn't we have caught this earlier?

Comment (coryan) [reply google/cloud/storage/internal/storage_round_robin.cc:291]: This is just defensive.  It really cannot (or should not) happen, but if it does then I would rather not crash and just stop using the invalid index.

Comment (devbww) [reply google/cloud/storage/internal/storage_round_robin.cc:291]: OK, thanks, although I've never been a big fan of that philosophy.

File: google/cloud/internal/async_connection_ready.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/internal/async_connection_ready.cc]: Removed

**Pull request merged**